mono/sgen/sgen-gray.c

```diff
@@ -146,14 +161,33 @@ sgen_gray_object_dequeue (SgenGrayQueue *queue)
 #endif
 
 	if (G_UNLIKELY (queue->cursor < GRAY_FIRST_CURSOR_POSITION (queue->first))) {
```

Comment:
This lock-free access to `queue->cursor` bothers me.  It's written to later down, too.  It's not clear at all to me that this is safe.